mozilla-services/socorro
# bug 1787931: fix super search fields to pull permissions from processed crash schema

This is a series of commits that move things around to eventually fix super search fields to pull permissions from the processed crash schema.

First, this removes `test_mapping` because it's dead code that's not used anymore. This fixes the `BreadcrumbsRule` and other things that used `get_schema()` and fixes `resolve_references` to return a new version of the data and not mutate the input. This fixes schema validation in `validate_processed_crash` so that it can be used for validating that the schemas themselves are valid in addition to testing them against processed crash data. This also moves some webapp-centric code from `socorro/external/es/super_search_fields.py` to `webapp-django/crashstats/supersearch/libsupersearch.py`. This also fixes the `upload_telemetry_schema` test because it prevented me from running tests using flags. :(

With all that cleanup done, then we add `default_permissions` to the schema so we're not defining defaults across the app.

Finally, we remove permissions from fields in super search fields and add some code to get the permissions from `processed_crash.schema.yaml`. We also add some code to the webapp to convert from schema permissions ("public", "protected", etc) to webapp permissions ("crashstats.view_pii"). This last bit of code is awkward feeling, but I didn't have a good way to generalize it and the webapp permissions are everywhere already.

To test this, I did this before the changes:

```
from socorro.external.es.super_search_fields import FIELDS


for key, val in FIELDS.items():
    print(f"{key}: {val['permissions_needed']}")
```

and this after:

```
from crashstats.supersearch.libsupersearch import SUPERSEARCH_FIELDS as FIELDS           

for key, val in FIELDS.items():                                                          
    print(f"{key}: {val['permissions_needed']}")    
```

And here's the differences:

```
> diff field_permissions_orig.txt field_permissions_new.txt 
39c39
< collector_notes: []
---
> collector_notes: ['crashstats.view_pii']
115c115
< remote_type: []
---
> remote_type: ['crashstats.view_pii']
119d118
< skunk_classification: []
124d122
< support_classification: []
```

The `*_classification` fields were removed because they don't refer to anything in the processed crash. The `remote_type` field had a typo in the `in_database_name` and fixing that fixes its permissions. `collector_notes` were marked protected in the `processed_crash.schema.yaml`, but not in `super_search_fields.py` which is a mistake.

I used `gmp_library_path` to verify that fields that are protected don't show up for people who don't have access to them.

I verified that *all fields* show up in the Super Search API reference.
